### PR TITLE
Single threaded K-Means training no longer uses a ForkJoinPool

### PR DESCRIPTION
**Pull request by Craigacp**
### Description
`KMeansTrainer` uses a `ForkJoinPool` to provide parallel computation of the e and m steps. This FJP is used even when operating in a single thread. The FJP requires extra permissions (namely "modifyThread") when running under a security manager, and in some security managers (notably OpenSearch's SecureSM) the FJP does not work at all due to a permissions issue inside OpenJDK. This means that the FJP breaks single threaded k-means training when running under a security manager, even though the FJP isn't necessary for sequential training. This change moves the FJP so it's only instantiated in multi-threaded execution paths, and also tidies up some of the code in `KMeansTrainer`. When using the FJP in `KMeansTrainer` and `KNNModel` it now uses a custom `ForkJoinWorkerThreadFactory` to allow appropriate security privileges to be granted to the FJP threads, so now parallel computation will work under `SecureSM` when given the appropriate security policy.

There's also a bug fix for the centroid computation when using weighted examples, as previously the weights were required to sum to 1 (but this is not guaranteed by the contract of the weights).

### Motivation
The FJP in `KMeansTrainer` interacts poorly with `SecureSM` used in OpenSearch as the security manager. This change means the FJP isn't used in sequential training allowing those runs to work under a `SecureSM`. With the custom `ForkJoinWorkerThreadFactory` now both `KMeansTrainer` and `KNNModel` should work under `SecureSM` with the appropriate security policy.

This issue was reported in #158.

### CODE REVIEW

File: Common/NearestNeighbour/src/main/java/org/tribuo/common/nearest/KNNModel.java [hunk 1]
```diff
@@ -495,4 +505,12 @@ public int compareTo(OutputDoublePair<T> o) {
         }
     }
 
+    /**
+     * Used to allow FJPs to work with OpenSearch's SecureSM.
+     */
+    private static final class CustomForkJoinWorkerThreadFactory implements ForkJoinPool.ForkJoinWorkerThreadFactory {
```
Comment (ylwu-amzn) [inline Common/NearestNeighbour/src/main/java/org/tribuo/common/nearest/KNNModel.java:511]: See same method defined in both `KMeansTrainer` and here, not sure if can define in some common place then these 2 classes can reuse directly. So we can just change one place if we need to tune this method in future.

Comment (Craigacp) [reply Common/NearestNeighbour/src/main/java/org/tribuo/common/nearest/KNNModel.java:511]: I agree it's not a good idea to have duplicated code, but I don't want to have the `CustomForkJoinWorkerThreadFactory` be a public class that becomes part of Tribuo's API. As Tribuo isn't using Java modules to guard implementation packages yet I think the options are make the class public or duplicate the code. If we end up requiring this in more than two places then we can spend some time and design a proper entry point (maybe a method on `org.tribuo.util.Util` which constructs a `ForkJoinPool` with the required parallelism).

Comment (ylwu-amzn) [reply Common/NearestNeighbour/src/main/java/org/tribuo/common/nearest/KNNModel.java:511]: Yeah, looks a factory/util class generates `ForkJoinPool` is one option, so we can avoid duplicate `CustomForkJoinWorkerThreadFactory` and `System.getSecurityManager() == null ? new ForkJoinPool(numThreads) : new ForkJoinPool(numThreads, THREAD_FACTORY, null, false);`

File: Clustering/KMeans/src/main/java/org/tribuo/clustering/kmeans/KMeansTrainer.java [hunk 2]
```diff
@@ -63,6 +67,10 @@
  * of threads used in the training step. The thread pool is local to an invocation of train,
  * so there can be multiple concurrent trainings.
  * <p>
+ * Note parallel training uses a {@link ForkJoinPool} which requires that the Tribuo codebase
+ * is given the "modifyThread" and "modifyThreadGroup" privileges when running under a
```
Comment (ylwu-amzn) [inline Clustering/KMeans/src/main/java/org/tribuo/clustering/kmeans/KMeansTrainer.java:71]: I think `modifyThreadGroup` is optional for `java.lang.SecurityManager`, only needed as OpenSearch `SecureSM` has stricter check of `modifyThreadGroup`

Comment (Craigacp) [reply Clustering/KMeans/src/main/java/org/tribuo/clustering/kmeans/KMeansTrainer.java:71]: So I agree, but the `modifyThreadGroup` check is triggered in the codepath, it's just that `java.lang.SecurityManager` doesn't care about it. So if there are other projects which use their own subclass of `SecurityManager` might need to care about it, and explaining that level of complexity didn't seem necessary, especially as the users of Tribuo might not understand the distinction.

File: Common/NearestNeighbour/src/main/java/org/tribuo/common/nearest/KNNModel.java (hunk 1 above)
Comment (Craigacp) [reply Common/NearestNeighbour/src/main/java/org/tribuo/common/nearest/KNNModel.java:511]: For the moment I'd like to leave it as is. It's all private so we can always change it later to point at a public endpoint when we have more pieces of code which need this functionality.

Comment (ylwu-amzn) [reply Common/NearestNeighbour/src/main/java/org/tribuo/common/nearest/KNNModel.java:511]: Sure, sound good to me

**Review by pogren — APPROVED** [review]
KMeansTrainer.train 
- conditionally instantiates a FJP if #threads is > 1.  
	- uses a FJP with a custom thread factory which uses the access controller - if a security manager is present
- cluster assignments are performed in a function called eStepFunc which can be used in parallel or in single-threaded mode.  It looks like a straightforward refactoring with no new logic (for the actual work that's being done)
- similar refactor of mStep  

KNNModel only updates the FJP to use the custom factory which uses the access controller.

**Pull request merged**